### PR TITLE
Random subsampling correction

### PR DESCRIPTION
**Pull request by yashpote**
Given N samples when K (N>K) samples are required, Barbarik should randomly choose the K from the set of N. However, the implementation currently does it in a way that may introduce bias. This pull request takes care of that issue.

### CODE REVIEW

**Comment by msoos** [comment]
Looks cool! Would it be okay if I review this on Sunday? Cheers,

Mate

File: barbarik.py [hunk 1]
```diff
@@ -171,10 +171,7 @@ def getSolutionFromUniGen(inputFile, numSolutions, indVarList, newSeed):
                 freq = int(line.split(':')[-1])
                 for i in range(freq):
                     solList.append(line.split(':')[0].replace('v', '').strip())
-                    if (len(solList) == numSolutions):
-                        break
-                if (len(solList) == numSolutions):
-                    break
```
Comment (msoos) [inline barbarik.py:177]: This seems wrong. We should be returning `numSolutions` here only.  Why do we need to change this?

File: barbarik.py [hunk 2]
```diff
@@ -834,11 +829,7 @@ def one_experiment(self, experiment, j, i, numExperiments, tj):
                 i += 1
                 ok, breakExperiment = exp.one_experiment(experiment, j, i, numExperiments, tj)
 
-                if ok is None:
```
Comment (msoos) [inline barbarik.py:837]: Wait, `None, None` is a valid return! Why is this being removed?

```
        if self.thresholdSolutions < self.minSamples:
            return None, None
```

File: barbarik.py (hunk 1 above)
Comment (yashpote) [reply barbarik.py:177]: Let's say we ask the sampler for 10 uniformly random solutions and it returns 20. The current implementation will only use the first 10 samples and discard the remaining 10. This is OK if the samples were returned in a randomised order. However, this is not ok if the sampler returned the samples in some sorted order.

The right thing to do in this case would be to randomly sample 10 out of the 20. This is exactly what we do in the very next line (178-180). 

AppMC3 and Unigen do actually return the samples in some sorted order, and thus cause the test to Reject incorrectly.

File: barbarik.py (hunk 2 above)
Comment (yashpote) [reply barbarik.py:837]: If ok is None or if ok is False, the correct thing to do is to continue to the next iteration. The real problem was with the 
`i -= 1` on line 841. The decrement is both theoretically unnecessary AND leads to an infinite loop as we don't change the value of seed before running the experiment again.

**Review by msoos — APPROVED** [review]

**Comment by msoos** [comment]
Just some minor feedback: we need to have tests. Currently, we don't have tests and this makes it VERY hard to reason about this program. It'd be nice to have some tests...

**Pull request merged**